cpp/src/tsne/bh_kernels.h

```diff
@@ -836,8 +807,56 @@ IntegrationKernel(const float eta,
 
     Y1[i] += ux;
     Y2[i] += uy;
+
+    // Confirm Y1 and Y2 are within bounds (-MAX_BOUNDS, MAX_BOUNDS)
+    // These are arbitrary but can help with outliers
+    // Also reset both gains and old forces
+    if (Y1[i] < -MAX_BOUNDS) {
+      Y1[i] = -MAX_BOUNDS;
+      gains1[i] = 1;
+      old_forces1[i] = 0;
+    }
+    else if (Y1[i] > MAX_BOUNDS) {
+      Y1[i] = MAX_BOUNDS;
+      gains1[i] = 1;
+      old_forces1[i] = 0;
+    }
+    if (Y2[i] < -MAX_BOUNDS) {
+      Y2[i] = -MAX_BOUNDS;
+      gains2[i] = 1;
+      old_forces2[i] = 0;
+    }
+    else if (Y2[i] > MAX_BOUNDS) {
+      Y2[i] = MAX_BOUNDS;
+      gains2[i] = 1;
+      old_forces2[i] = 0;
+    }
+
+    atomicAdd(&sums[0], Y1[i]);
+    atomicAdd(&sums[1], Y2[i]);
+  }
+}
+
+
+__global__ void
+mean_centre(float *restrict Y1,
+            float *restrict Y2,
+            const float *restrict means,
+            const float N)
+{
+  const int i = (blockIdx.x * blockDim.x) + threadIdx.x;
+  if (i >= N) return;
+  Y1[i] -= means[0];
```

Comment:
While it might require more kernels to be executed in the end, I think using unary op prims for this would be better

Reply:
Good point

Reply:
Oh wait I just realized part of it also adds some jitter noise to every 1000th element. I'm not exactly sure to do it with the cuML prims :(